scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,444 @@
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let checkpoint = receiver.borrow().clone();
+
+            checkpoint_saver.save_checkpoint(checkpoint).await?;
+
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for saving checkpoints.
+#[async_trait]
+pub trait CDCCheckpointSaver: Send + Sync {
+    /// Saves given checkpoint.
+    async fn save_checkpoint(&self, checkpoint: Checkpoint) -> anyhow::Result<()>;
+    /// Loads last seen generation from the checkpoint table.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    /// Loads last seen timestamp for given stream id from the checkpoint table.
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+
+/// Default implementation for [`CDCCheckpointSaver`] trait.
+pub struct CDCDefaultCheckpointSaver {
+    session: Arc<Session>,
+    checkpoint_table: String,
+    make_checkpoint_stmt: PreparedStatement,
+    last_seen_generation: Arc<Mutex<GenerationTimestamp>>,
+}
+
+impl CDCDefaultCheckpointSaver {
+    /// Creates new [`CDCDefaultCheckpointSaver`].
+    /// If given `None` value as table name,
+    /// it uses default value `"CDC_CHECKPOINT"`.
+    pub async fn new(
+        session: &Arc<Session>,
+        keyspace: &str,
+        tablename: Option<&str>,
+    ) -> anyhow::Result<Self> {
+        let session = Arc::clone(session);
+        let tablename = tablename.unwrap_or("CDC_CHECKPOINT");
+        let checkpoint_table = format!("{}.{}", keyspace, tablename);
+
+        CDCDefaultCheckpointSaver::create_checkpoints_table(&session, &checkpoint_table).await?;
+
+        let make_checkpoint_stmt = session
+            .prepare(format!(
+                "UPDATE {} SET generation = ?, time = ?
+                WHERE stream_id = ?",
+                checkpoint_table
+            ))
+            .await?;
+
+        let cp_saver = CDCDefaultCheckpointSaver {
+            session,
+            checkpoint_table,
+            make_checkpoint_stmt,
+            last_seen_generation: Arc::new(Mutex::new(GenerationTimestamp {
+                timestamp: chrono::Duration::min_value(),
+            })),
+        };
+
+        Ok(cp_saver)
+    }
+
+    // Creates new table with specified schema in the given keyspace with provided name.
+    // If a table with such name already exists, doesn't do anything.
+    async fn create_checkpoints_table(
+        session: &Arc<Session>,
+        checkpoint_table: &str,
+    ) -> anyhow::Result<()> {
+        let schema = get_checkpoint_table_schema(checkpoint_table);
+
+        session.query(schema, ()).await?;
+        session.await_schema_agreement().await?;
+
+        Ok(())
+    }
+}
+
+fn get_default_generation_pk() -> Vec<u8> {
+    vec![0]
```

Comment:
Why not just use an empty vec? IMO it differentiates itself better from other generations (empty vs non-empty) and should resolve that one issue with adding `42` to the stream number.

Reply:
I have changed return type to `StreamId` as I thought it would be more elegant than using bare vector and returning an empty stream id results in: `Database returned an error: The query is syntatically correct but invalid, Error message: Key may not be empty'`.